applications/GeoMechanicsApplication/custom_utilities/convergence_criteria_factory.hpp

```diff
@@ -26,25 +28,32 @@ class ConvergenceCriteriaFactory
     static std::shared_ptr<ConvergenceCriteriaType> Create(const Parameters& rSolverSettings)
     {
         KRATOS_ERROR_IF_NOT(rSolverSettings.Has("convergence_criterion"))
-        << "No convergence_criterion is defined, aborting.";
+            << "No convergence_criterion is defined, aborting.";
 
-        if (rSolverSettings["convergence_criterion"].GetString() == "displacement_criterion")
+        if (rSolverSettings["convergence_criterion"].GetString() ==
+            "displacement_criterion")
         {
-            Parameters convergenceInputs;
-
-            std::vector<std::string> entries_to_copy = {"displacement_absolute_tolerance", "displacement_relative_tolerance"};
-            for (const std::string& entry : entries_to_copy)
-            {
-                if (rSolverSettings.Has(entry))
-                {
-                    convergenceInputs.AddValue(entry, rSolverSettings[entry]);
-                }
-            }
-            return std::make_shared<DisplacementCriteria<TSparseSpace, TDenseSpace>>(convergenceInputs);
+            const std::vector<std::string> entries_to_copy = {
+                "displacement_absolute_tolerance",
+                "displacement_relative_tolerance"};
+            Parameters convergence_inputs = ParametersUtilities::ExtractParameters(rSolverSettings, entries_to_copy);
+            return std::make_shared<DisplacementCriteria<TSparseSpace, TDenseSpace>>(convergence_inputs);
+        }
+        if (rSolverSettings["convergence_criterion"].GetString() ==
+            "residual_criterion")
+        {
+            const std::vector<std::string> entries_to_copy = {
+                "residual_absolute_tolerance", "residual_relative_tolerance"};
+            Parameters convergence_inputs = ParametersUtilities::ExtractParameters(rSolverSettings, entries_to_copy);
+            return std::make_shared<ResidualCriteria<TSparseSpace, TDenseSpace>>(convergence_inputs);
         }
```

Comment:
Correct functionality as far as I can read. Still the 2 branches are very similar, the difference being displacement or residual in the names. I wonder if it is worth to make that generic, such that water_pressure and temperature can be done in future without making another copy.

Reply:
Good point, that should be possible, let's keep it in mind when we start adding the rest!

Reply:
I agree with @WPK4FEM's comment. We could consider to follow the same strategy as we did for the process factory, where the factory itself is very generic and creators can be added as needed. But for me, it is fine to do that in a separate PR.